homeassistant/helpers/script.py

```diff
@@ -223,288 +666,59 @@ def referenced_entities(self):
     def run(self, variables=None, context=None):
         """Run script."""
         asyncio.run_coroutine_threadsafe(
-            self.async_run(variables, context), self.hass.loop
+            self.async_run(variables, context), self._hass.loop
         ).result()
 
     async def async_run(
         self, variables: Optional[Sequence] = None, context: Optional[Context] = None
-    ) -> None:
-        """Run script.
-
-        This method is a coroutine.
-        """
-        self.last_triggered = date_util.utcnow()
-        if self._cur == -1:
-            self._log("Running script")
-            self._cur = 0
-
-        # Unregister callback if we were in a delay or wait but turn on is
-        # called again. In that case we just continue execution.
-        self._async_remove_listener()
-
-        for cur, action in islice(enumerate(self.sequence), self._cur, None):
-            try:
-                await self._handle_action(action, variables, context)
-            except _SuspendScript:
-                # Store next step to take and notify change listeners
-                self._cur = cur + 1
-                if self._change_listener:
-                    self.hass.async_add_job(self._change_listener)
-                return
-            except _StopScript:
-                break
-            except Exception:
-                # Store the step that had an exception
-                self._exception_step = cur
-                # Set script to not running
-                self._cur = -1
-                self.last_action = None
-                # Pass exception on.
-                raise
-
-        # Set script to not-running.
-        self._cur = -1
-        self.last_action = None
-        if self._change_listener:
-            self.hass.async_add_job(self._change_listener)
+    ) -> bool:
+        """Run script."""
+        if self.is_running:
+            if self._if_running == IF_RUNNING_IGNORE:
+                self._log("Skipping script")
+                return False
+
+            if self._if_running == IF_RUNNING_ERROR:
+                self._raise("Already running")
+            if self._if_running == IF_RUNNING_RESTART:
+                self._log("Restarting script")
+                self.async_stop()
```

Comment:
Should we await this?

Reply:
```suggestion
                await self.async_stop()
```

Reply:
For sure. Missed that one. As you can probably tell, I haven't run many tests on the new code yet. I've done some ad hoc testing, and have written a few new tests, but I was waiting until the code settled down a bit before getting serious about testing the new code. 😅